diff --git a/src/qemu/qemu_conf.c b/src/qemu/qemu_conf.c
index 7c63bdc..13980b1 100644
--- a/src/qemu/qemu_conf.c
+++ b/src/qemu/qemu_conf.c
@@ -3707,9 +3707,17 @@ int qemudBuildCommandLine(virConnectPtr conn,
      * 2. The qemu binary has the -no-kvm flag
      */
     if ((qemuCmdFlags & QEMUD_CMD_FLAG_KVM) &&
-        def->virtType == VIR_DOMAIN_VIRT_QEMU)
+        def->virtType == VIR_DOMAIN_VIRT_QEMU) {
         disableKVM = 1;
 
+        /*
+         * do not use boot=on for drives when not using KVM since this
+         * is not supported at all in upstream QEmu.
+         */
+        if (qemuCmdFlags & QEMUD_CMD_FLAG_DRIVE_BOOT)
+            qemuCmdFlags -= QEMUD_CMD_FLAG_DRIVE_BOOT;
+    }
+
     /* Should explicitly enable KVM if
      * 1. Guest domain is 'kvm'
      * 2. The qemu binary has the -enable-kvm flag
